crates/lib3h/src/transport/transport_encoding/mod.rs

```diff
@@ -0,0 +1,641 @@
+    }
+
+    impl TransportMock {
+        pub fn new(
+            mock_sender: crossbeam_channel::Sender<(Url, Vec<u8>)>,
+            mock_receiver: crossbeam_channel::Receiver<(Url, Vec<u8>)>,
+        ) -> Self {
+            let (endpoint_parent, endpoint_self) = create_ghost_channel();
+            let endpoint_parent = Some(endpoint_parent);
+            let endpoint_self = Detach::new(endpoint_self.as_context_endpoint("mock_to_parent_"));
+            Self {
+                endpoint_parent,
+                endpoint_self,
+                bound_url: Url::parse("none:").expect("can parse url"),
+                mock_sender,
+                mock_receiver,
+            }
+        }
+    }
+
+    impl
+        GhostActor<
+            RequestToParent,
+            RequestToParentResponse,
+            RequestToChild,
+            RequestToChildResponse,
+            TransportError,
+        > for TransportMock
+    {
+        fn as_any(&mut self) -> &mut dyn Any {
+            &mut *self
+        }
+
+        fn take_parent_endpoint(&mut self) -> Option<TransportActorParentEndpoint> {
+            std::mem::replace(&mut self.endpoint_parent, None)
+        }
+
+        fn process_concrete(&mut self) -> GhostResult<WorkWasDone> {
+            detach_run!(&mut self.endpoint_self, |es| es.process(self.as_any()))?;
+            for mut msg in self.endpoint_self.as_mut().drain_messages() {
+                match msg.take_message().expect("exists") {
+                    RequestToChild::Bind { mut spec } => {
+                        spec.set_path("bound");
+                        self.bound_url = spec.clone();
+                        msg.respond(Ok(RequestToChildResponse::Bind(BindResultData {
+                            bound_url: spec,
+                        })));
+                    }
+                    RequestToChild::SendMessage { address, payload } => {
+                        self.mock_sender.send((address, payload)).unwrap();
+                        msg.respond(Ok(RequestToChildResponse::SendMessage));
+                    }
+                }
+            }
+            loop {
+                match self.mock_receiver.try_recv() {
+                    Ok((address, payload)) => {
+                        // bit of a hack, just always send an incoming connection
+                        // in front of all received data messages
+                        self.endpoint_self
+                            .publish(RequestToParent::IncomingConnection {
+                                address: address.clone(),
+                            });
+                        self.endpoint_self
+                            .publish(RequestToParent::ReceivedData { address, payload });
+                    }
+                    Err(_) => break,
+                }
+            }
+            Ok(false.into())
+        }
+    }
+
+    #[test]
+    fn it_should_exchange_messages() {
```

Comment:
I'll see what I can do to make this tests less fragile eventually.

Reply:
sweet :+1: